File: Part1_강의자료/Ch08_투포인터/문제별코드/10025_게으른백곰/10025.java
import java.util.Arrays;
import java.util.Scanner;

class Main
{
    static class Bucket {
        int g;
        int x;
        Bucket(int g, int x) {
            this.g = g;
            this.x = x;
        }
    }

    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int K = sc.nextInt();
        Bucket[] buckets = new Bucket[N];
        for (int i = 0; i < N; i++)
            buckets[i] = new Bucket(sc.nextInt(), sc.nextInt());

        Arrays.sort(buckets, (o1, o2) -> o1.x - o2.x);

        int nextIndex = 0;
        int maxSum = 0;
        int currentSum = 0;
        for (int i = 0; i < N; i++) {
            while (nextIndex < N && buckets[nextIndex].x - buckets[i].x <= 2 * K) {
                currentSum += buckets[nextIndex++].g;
            }
            maxSum = Math.max(maxSum, currentSum);
            currentSum -= buckets[i].g;
        }
        System.out.println(maxSum);
    }
}

File: Part1_강의자료/Ch08_투포인터/문제별코드/10025_게으른백곰/10025.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

struct Bucket {
    int g;
    int x;
    bool operator<(const Bucket &o)const {
        return x < o.x;
    }
};

int main()
{
    int N, K;

    scanf("%d %d", &N, &K);
    vector<Bucket> buckets(N);
    for (int i = 0; i < N; i++)
        scanf("%d %d", &buckets[i].g, &buckets[i].x);

    sort(buckets.begin(), buckets.end());

    int next = 0;
    int max_sum = 0;
    int cur_sum = 0;
    for (int i = 0; i < N; i++)
    {
        while (next < N && buckets[next].x - buckets[i].x <= 2 * K)
            cur_sum += buckets[next++].g;
        max_sum = max(max_sum, cur_sum);
        cur_sum -= buckets[i].g;
    }

    printf("%d\n", max_sum);
    return 0;
}
